### ISSUE TITLE
SetDeadline not working on ptmx

### ISSUE THREAD

**Issue by gcrtnst** [creack/pty#162]
Run the following code:
```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/creack/pty"
)

func main() {
	ptmx, pts, err := pty.Open()
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: open: %v\n", err)
		return
	}
	defer ptmx.Close()
	defer pts.Close()

	err = ptmx.SetDeadline(time.Now().Add(5 * time.Second))
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: set deadline: %v\n", err)
		return
	}

	n, err := ptmx.Read(make([]byte, 1))
	fmt.Printf("n = %d\n", n)
	fmt.Printf("err = %v\n", err)
}
```

Expected behavior: The code exits after 5 seconds and the output following
```
n = 0
err = read /dev/ptmx: i/o timeout
```

Actual behavior: The code does not exit forever, it blocks on the `ptmx.Read` call and does not output anything to the stdout/stderr.

The cause of the issue is that "github.com/creack/pty" package calls `(*os.File).Fd` internally, which [disables the SetDeadline method on Unix systems](https://pkg.go.dev/os#File.Fd:~:text=On%20Unix%20systems%20this%20will%20cause%20the%20SetDeadline%20methods%20to%20stop%20working.). Using `(*os.File).SyscallConn` instead may solve this issue.

[This other issue posted on StackOverflow](https://stackoverflow.com/questions/72721691/pty-doesnt-get-closed-althoguh-close-syscall-is-called) may have the same cause.

I reproduced this issue on Linux, but it may happen on other platforms.

**Comment by sio** [comment]
@gcrtnst, thank you for the detailed report! I have encountered a similar problem with blocking Read() and your mention of side effects in `(*os.File).Fd()` had saved me a significant amount of time.

I have submitted a PR (#167) to fix this problem, but until it gets merged you may be interested in a workaround. You can manually reset file descriptor to non-blocking mode:

```diff
--- issue162/demo-fail.go
+++ issue162/demo-ok.go
@@ -3,6 +3,7 @@
 import (
 	"fmt"
 	"os"
+	"syscall"
 	"time"
 
 	"github.com/creack/pty"
@@ -17,6 +18,11 @@
 	defer ptmx.Close()
 	defer pts.Close()
 
+	err = syscall.SetNonblock(int(ptmx.Fd()), true)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to unblock ptmx: %v\n", err)
+	}
+
 	err = ptmx.SetDeadline(time.Now().Add(5 * time.Second))
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "error: set deadline: %v\n", err)
```

Of course, this will only work until you make another call to `Fd()` through pty library, though it's often good enough for my use cases.

**Comment by creack** [comment]
Closed with #167
